### PR TITLE
fix: create shards without overlaps

### LINKED ISSUE #22598: Backport to 1.x: Overlapping shard fix
Backports https://github.com/influxdata/influxdb/issues/22596 to 1.x

### PR DESCRIPTION
**Pull request by lesam**
Closes #22598 

- [x] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
- [x] Rebased/mergeable
- [x] Tests pass

### CODE REVIEW

File: services/meta/data.go [hunk 1]
```diff
@@ -369,16 +369,48 @@ func (data *Data) CreateShardGroup(database, policy string, timestamp time.Time)
...
+		// Assume the above conditions are true for shards index < i, we want to modify startTime,endTime so they are true
+		// also for shard_i
+
+		// It must be the case that either endI <= timestamp || timestamp < startI, because otherwise:
+		// startI <= timestamp < endI means timestamp is contained in shard I
+		if !timestamp.Before(endI) && endI.After(startTime) {
+			// startTime < endI <= timestamp
+			startTime = endI
+		}
+		if startI.After(timestamp) && startI.Before(endTime) {
```
Comment (davidby-influx) [inline services/meta/data.go:402]: I sort of would like sanity checks that our invariants are actually true.  Life if we run this code on files which have overlapping shards, for instance.  Not a critical wish, just a secondary verification of sanity for the existing shards.

**Review by davidby-influx — APPROVED** [review]
Non-blocking wish list for invariant checking....

File: services/meta/data.go (hunk 1 above)
Comment (lesam) [reply services/meta/data.go:402]: We actually already check above that the timestamp we want to add is not contained in an existing shard: 

```
if rpi.ShardGroupByTimestamp(timestamp) != nil {
	return nil
}
```

So we really just walked all the existing shards and validated that there is no shard with `startI <= timestamp < endI`.

And the algo here is robust against existing overlapping shards - we just find the maximum region:
- contained within the original [start, end)
- also containing timestamp
- that doesn't overlap with any existing shards

So if there are existing overlaps we don't fix them, but we don't create any new overlaps.

Comment (davidby-influx) [reply services/meta/data.go:402]: I was thinking we should flag them with a log message.

Comment (lesam) [reply services/meta/data.go:402]: Hmm - I thought that would be a bit noisy for all the customers with existing overlaps, but maybe that is a good idea...

Comment (davidby-influx) [reply services/meta/data.go:402]: Only at Debug level?

File: services/meta/data.go [hunk 2]
```diff
@@ -369,16 +369,48 @@ func (data *Data) CreateShardGroup(database, policy string, timestamp time.Time)
...
+
+	for i := range rpi.ShardGroups {
+		if rpi.ShardGroups[i].Deleted() {
+			continue
+		}
+		startI := rpi.ShardGroups[i].StartTime
+		endI := rpi.ShardGroups[i].EndTime
+		if rpi.ShardGroups[i].Truncated() {
+			endI = rpi.ShardGroups[i].TruncatedAt
+		}
```
Comment (gwossum) [inline services/meta/data.go:387]: This basic code is copied and pasted a few times in the code. Would it make sense to add an `ShardGroupInfo.ActualEndTime()` method? I have a fear we'll find a bug one day due to one section of code not adjusting for `.TruncatedAt`

**Pull request merged**